syntax = "proto3";

package serializationPackage;

message VMData {

  repeated int32 instructions = 1;
  map<int64, ValueType> constantVals = 2;
  map<string, AddressAndHash> stringsAtAddresses = 3;

  message AddressAndHash{
    int64 address = 1;
    int32 hash = 2;
  }

  enum Opcode
  {
      OP_CONSTANT = 0;
      OP_NIL = 1;
      OP_TRUE = 2;
      OP_FALSE = 3;
      OP_POP = 4;
      OP_GET_LOCAL = 5;
      OP_SET_LOCAL = 6;
      OP_GET_GLOBAL = 7;
      OP_DEFINE_GLOBAL = 8;
      OP_SET_GLOBAL = 9;
      OP_EQUAL = 10;
      OP_GREATER = 11;
      OP_LESS = 12;
      OP_ADD = 13;
      OP_SUBTRACT = 14;
      OP_MULTIPLY = 15;
      OP_DIVIDE = 16;
      OP_NOT = 17;
      OP_NEGATE = 18;
      OP_PRINT = 19;
      OP_JUMP = 20;
      OP_JUMP_IF_FALSE = 21;
      OP_LOOP = 22;
      OP_RETURN = 23;
  } 
}

message ValueType {
  oneof ValueTypes{
    int32 objectAddress = 1;
    double numVal = 2;
    bool boolVal = 4;
  }
}



